// metals-languageclient/lib/interfaces/TreeViewProtocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetalsTreeViewReveal = exports.MetalsTreeViewNodeCollapseDidChange = exports.MetalsTreeViewVisibilityDidChange = exports.MetalsTreeViewParent = exports.MetalsTreeViewDidChange = exports.MetalsTreeViewChildren = void 0;
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
var MetalsTreeViewChildren;
(function (MetalsTreeViewChildren) {
    MetalsTreeViewChildren.type = new vscode_languageserver_protocol_1.RequestType("metals/treeViewChildren");
})(MetalsTreeViewChildren = exports.MetalsTreeViewChildren || (exports.MetalsTreeViewChildren = {}));
var MetalsTreeViewDidChange;
(function (MetalsTreeViewDidChange) {
    MetalsTreeViewDidChange.type = new vscode_languageserver_protocol_1.NotificationType("metals/treeViewDidChange");
})(MetalsTreeViewDidChange = exports.MetalsTreeViewDidChange || (exports.MetalsTreeViewDidChange = {}));
var MetalsTreeViewParent;
(function (MetalsTreeViewParent) {
    MetalsTreeViewParent.type = new vscode_languageserver_protocol_1.RequestType("metals/treeViewParent");
})(MetalsTreeViewParent = exports.MetalsTreeViewParent || (exports.MetalsTreeViewParent = {}));
var MetalsTreeViewVisibilityDidChange;
(function (MetalsTreeViewVisibilityDidChange) {
    MetalsTreeViewVisibilityDidChange.type = new vscode_languageserver_protocol_1.NotificationType("metals/treeViewVisibilityDidChange");
})(MetalsTreeViewVisibilityDidChange = exports.MetalsTreeViewVisibilityDidChange || (exports.MetalsTreeViewVisibilityDidChange = {}));
var MetalsTreeViewNodeCollapseDidChange;
(function (MetalsTreeViewNodeCollapseDidChange) {
    MetalsTreeViewNodeCollapseDidChange.type = new vscode_languageserver_protocol_1.NotificationType("metals/treeViewNodeCollapseDidChange");
})(MetalsTreeViewNodeCollapseDidChange = exports.MetalsTreeViewNodeCollapseDidChange || (exports.MetalsTreeViewNodeCollapseDidChange = {}));
var MetalsTreeViewReveal;
(function (MetalsTreeViewReveal) {
    MetalsTreeViewReveal.type = new vscode_languageserver_protocol_1.RequestType("metals/treeViewReveal");
})(MetalsTreeViewReveal = exports.MetalsTreeViewReveal || (exports.MetalsTreeViewReveal = {}));
